### ISSUE TITLE
Implicit line continuation before "Implements"

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/vblang#232]
Lines that contains the word Implements in Method detentions as too long. It will be easier to consider  this word an implicit line continuation position. 

```
Public Iterator Function GetEnumerator() As IEnumerator
           Implements IEnumerable.GetEnumerator

```
This will be consistent with the class definition where the Impliments statements starts at a new line.

**Comment by AnthonyDGreen** [comment]
Agreed. Handles too! Approved.

**Comment by ericmutta** [comment]
This would be really handy! And while we are at it, **lets generically say implicit line continuations are allowed around ANY keyword that appears in a function/sub declaration, not just the Implements one**.

Consider the case with tuples which have a tendency to make lines really really long:

```vb
  Public Interface IFoo
    Function GetStuff() As (Foo As String, Bar As Integer, Baz As ExecutionEngineException)
  End Interface

  Public Class CFoo
    Implements IFoo

    Public Function GetStuff() As (Foo As String, Bar As Integer, Baz As ExecutionEngineException) Implements IFoo.GetStuff
      Throw New NotImplementedException()
    End Function
  End Class
```

With implicit line continuations around any keyword in a function/sub declaration, that `GetStuff` method in class `CFoo` could be written like this (with the editor doing the right indenting when you press Enter):

```vb
    Public Function GetStuff() 
           As (Foo As String, Bar As Integer, Baz As ExecutionEngineException) 
           Implements IFoo.GetStuff

      Throw New NotImplementedException()
    End Function
```

The above has implicit line continuations around the `As` and `Implements` keywords. Since `Function` is also a keyword, you could also do this (handy for people who like really descriptive method names):

```vb
    Public Function 
           OneReallyLongAndDescriptiveMethodNameHere() 
           As (Foo As String, Bar As Integer, Baz As ExecutionEngineException) 
           Implements IFoo.GetStuff

      Throw New NotImplementedException()
    End Function
```

Implicit line continuations are one of those small things that have a really big productivity boost and give VB more flexibility in formatting code.  I hope to see them in more places!

**Comment by MohammadHamdyGhanem** [comment]
@AnthonyDGreen 
Same goes for From in collection intialization.

**Comment by reduckted** [comment]
Is it possible to get rid of line continuation markers everywhere? Is there anything that would definitely _require_ a line continuation market to be parsed unambiguously?

**Comment by reduckted** [comment]
Hmm, even with a `Return` statement, you'd know whether something _should_ follow it based on whether you're in a `Function` or a `Sub`.

**Comment by Echo-8-ERA** [comment]
I do sometimes use the old style of returning values from functions to avoid having to declare a `ret` variable. If it were ever allowed to use a `Return` without return value (instead of using `Exit Function`) in functions for that case, it would break. It doesn't work though, so I guess it doesn't matter. 

Well, the other one I thought of was`End`, but I'm not sure if it still terminated the application like in classic VB.

**Comment by paul1956** [comment]
Is there any way to make it work before a . like in C#, so the next like can start with the .?

**Comment by reduckted** [comment]
Yeah, I'd love that. I guess there might be some difficulty working out what the intent is inside a `With` block though.

**Comment by franzalex** [comment]
Maybe the `With` block can be treated differently from the other cases. Specifically, once the parser encounters a `With` block, then it will interpret lines tarting with the `.` like it does now if the identifier after the `.` is not a member of the item initialised in the `With` block.

**Comment by ericmutta** [comment]
> @paul1956 Is there any way to make it work before a . like in C#, so the next like can start with the .?

That would be pretty handy, especially if the behaviour was extended to all operators so we could do things like this:

```vb
        Dim SomeStringValue = "foo"
                            & "bar"  '<--- note the & operator was preceded by CRLF
                            & "baz"
```

My personal favourite would be supporting that behaviour for the comma seperator (I miss this ability from T-SQL's very flexible syntax):

```vb
    SomeFunction(
       Value1
      ,Value2  '<--- note the comma appears BEFORE the argument.
      ,Value3
    )
```

PS: happy new year everyone!

**Comment by MohammadHamdyGhanem** [comment]
@AnthonyDGreen 
It will be useful if you add implicit line continuation before +, -, *, \, /, And, Or, Not, ....etc
I have this code in C#
```
var email = WordEdge
        + ValidFirstChar
        + ValidChars[0, 19]
        + "@"
        + ValidFirstChar
        + ValidChars[2, 14]
        + "."
        + ValidFirstChar
        + ValidChars[1, 9]
        + WordEdge;
```
Similar code doesn't work in VB.NET although no new line can start with + in VB.NET:
```
Dim email = WordEdge
        + ValidFirstChar
        + ValidChars(0, 19)
        + "@"
        + ValidFirstChar
        + ValidChars(2, 14)
        + "."
        + ValidFirstChar
        + ValidChars(1, 9)
        + WordEdge
```

**Comment by paul1956** [comment]
I ran into a related issue with multiple Attributes where the compiler wants _ after attribuate1 but actually does allow them. Lines with only comments, blank lines and # directives should be treated as if they don't exist when parsing or I should be allowed to use a line continuation to say I'm not done.

```
<attribuate1>
' comment
<Fact>

**Comment by KathleenDollard** [comment]
When implicit line continuation was added in some cases, there were niche case concerns. I don't think the places proposed here are going to run into that, but I don't think we can abandon all forced line continuation. 

<Joke>
' We could always go to recognizing end of line by the number of spaces/indent!
<EndJoke>

**Comment by jrmoreno1** [comment]
Implicit continuation after a { would also be nice -- I just turned { _ into {( inside a string interpolation, because I consider it slightly more readable, but neither is really good, also before a }..

**Comment by paul1956** [comment]
@gafter while I would like comment VB needs directives between attributes below is legal in C# but not VB
```
    <Serializable>
#If Not FEATURE_CORECLR Then
    <SecurityPermissionAttribute(SecurityAction.InheritanceDemand, ControlEvidence:=True, ControlPolicy:=True)>
#End If
    <System.Runtime.InteropServices.ComVisible(True)>
    Public MustInherit Class CodeAccessPermission
        Implements IPermission, ISecurityEncodable, IStackWalk
```
 and the workaround requires duplicating too much code
```
#If Not FEATURE_CORECLR Then
    <Serializable>
    <SecurityPermissionAttribute(SecurityAction.InheritanceDemand, ControlEvidence:=True, ControlPolicy:=True)>
    <System.Runtime.InteropServices.ComVisible(True)>
    Public MustInherit Class CodeAccessPermission
        Implements IPermission, ISecurityEncodable, IStackWalk
#Else
    <Serializable>
    <System.Runtime.InteropServices.ComVisible(True)>
    Public MustInherit Class CodeAccessPermission
        Implements IPermission, ISecurityEncodable, IStackWalk
#End If
```

**Comment by AnthonyDGreen** [comment]
@paul1956 that's in the same vein as #65 in that it puts more work on the IDE to properly handle "significant" trivia between tokens. But the IDE team has said they already do this for C# so it's not the end of the world and they're not sure they rely on that "invariant" much anyway. I think it's a separate proposal from this and #65 though it might be done at the same time as #65.

**Comment by jrmoreno1** [comment]
The roslyn request https://github.com/dotnet/roslyn/issues/4885, was a precursor to this

**Comment by paul1956** [comment]
@AnthonyDGreen  Adding comments after _ is relative simple, testing is a lot of work. IDE was almost untouched, just the SmartIndenter required a minor fix. I think #65 is ready for real review. You can now put comment between Attributes but not on the next line

```
       <ExcludeFromCodeCoverage> _ '  Test
        <Extension>
```

**Comment by AnthonyDGreen** [comment]
@paul1956 link to a branch or PR?

**Comment by paul1956** [comment]
@AnthonyDGreen  https://github.com/dotnet/roslyn/pull/28327

**Comment by jrmoreno1** [comment]
@gafter : could you pull 4885 over into vblang?  Or does someone have to reenter it by hand?

**Comment by gafter** [comment]
Transferred to https://github.com/dotnet/vblang/issues/489
